File: akf_projects/akf_projects/page/project_survey_dashboard/project_survey_dashboard.py
#main-dashboard
import frappe
from frappe import _
from frappe.utils import getdate, add_days, nowdate


@frappe.whitelist(allow_guest=True)
def get_information(project=None):
    return {
        "region_wise_survey": get_region_wise_survey(project),
        "district_wise_survey": get_district_wise_survey(project),
        "tehsil_wise_survey": get_tehsil_wise_survey(project),
        "product_wise_survey": get_product_wise_survey(project),
        "allocated_vs_unallocated_survey": get_allocated_vs_unallocated_survey(project),
        "approved_vs_unapproved_survey": get_approved_vs_unapproved_survey(project),
        "survey_forms_count": get_total_survey_forms(project),
    }

def get_region_wise_survey(project=None):
    query = """
        SELECT 
            CASE
                WHEN region IS NULL OR region = '' THEN 'N/A'
                ELSE region
            END AS region,
            COUNT(name) as survey_count
        FROM `tabProject Survey Forms`
        GROUP BY 
            CASE
                WHEN region IS NULL OR region = '' THEN 'N/A'
                ELSE region
            END      
    """
    result = frappe.db.sql(query, as_dict=False)
    # frappe.msgprint("Region:")
    # frappe.msgprint(frappe.as_json(result))
    
    # total_surveys = sum(row[1] for row in result)
    total_surveys = 0
    return {"data": result, "total_surveys": total_surveys}

def get_district_wise_survey(project=None):
    query = """
        SELECT 
            CASE
                WHEN district IS NULL OR district = '' THEN 'N/A'
                ELSE district
            END AS district,
            COUNT(name) as survey_count
        FROM `tabProject Survey Forms`
        GROUP BY 
            CASE
                WHEN district IS NULL OR district = '' THEN 'N/A'
                ELSE district
            END      
    """
    result = frappe.db.sql(query, as_dict=False)
    # frappe.msgprint("district:")
    # frappe.msgprint(frappe.as_json(result))
    
    # total_surveys = sum(row[1] for row in result)
    total_surveys = 0
    return {"data": result, "total_surveys": total_surveys}

def get_tehsil_wise_survey(project=None):
    query = """
        SELECT 
            CASE
                WHEN tehsil IS NULL OR tehsil = '' THEN 'N/A'
                ELSE tehsil
            END AS tehsil,
            COUNT(name) as survey_count
        FROM `tabProject Survey Forms`
        GROUP BY 
            CASE
                WHEN tehsil IS NULL OR tehsil = '' THEN 'N/A'
                ELSE tehsil
            END      
    """
    result = frappe.db.sql(query, as_dict=False)
    # frappe.msgprint("tehsil:")
    # frappe.msgprint(frappe.as_json(result))
    
    # total_surveys = sum(row[1] for row in result)
    total_surveys = 0
    return {"data": result, "total_surveys": total_surveys}

def get_product_wise_survey(project=None):
    query = """
        SELECT 
            CASE
                WHEN product IS NULL OR product = '' THEN 'N/A'
                ELSE product
            END AS product,
            COUNT(name) as survey_count
        FROM `tabProject Survey Forms`
        GROUP BY 
            CASE
                WHEN product IS NULL OR product = '' THEN 'N/A'
                ELSE product
            END      
    """
    result = frappe.db.sql(query, as_dict=False)
    # frappe.msgprint("product:")
    # frappe.msgprint(frappe.as_json(result))
    
    # total_surveys = sum(row[1] for row in result)
    total_surveys = 0
    return {"data": result, "total_surveys": total_surveys}

def get_allocated_vs_unallocated_survey(project=None):
    query = """
        SELECT 
            COUNT(CASE WHEN allocation_status = 'Allocated' THEN 1 END) AS allocated,
            COUNT(CASE WHEN allocation_status = 'Unallocated' THEN 1 END) AS unallocated
        FROM `tabProject Survey Forms`
    """
    result = frappe.db.sql(query, as_dict=False)
    
    allocated = result[0][0] if result else 0
    unallocated = result[0][1] if result else 0

    
    return {
        "allocated": allocated,
        "unallocated": unallocated
    }

def get_approved_vs_unapproved_survey(project=None):
    query = """
        SELECT 
            COUNT(CASE WHEN approval_status = 'Pending' THEN 1 END) AS pending,
            COUNT(CASE WHEN approval_status = 'Verified By Regional' THEN 1 END) AS verified,
            COUNT(CASE WHEN approval_status = 'Approved By Head Office' THEN 1 END) AS approved
        FROM `tabProject Survey Forms`
    """
    result = frappe.db.sql(query, as_dict=False)
    
    pending = result[0][0] if result else 0
    verified = result[0][1] if result else 0
    approved = result[0][2] if result else 0

    
    return {
        "pending": pending,
        "verified": verified,
        "approved": approved
    }

def get_total_survey_forms(project=None):
    query = """
        SELECT 
            COUNT(name) as survey_count
        FROM `tabProject Survey Forms`
    """
    result = frappe.db.sql(query, as_dict=False)
    
    return result[0][0] if result else 0

File: akf_projects/akf_projects/page/project_survey_dashboard/project_survey_dashboard.js
// let filters = {};
// frappe.pages['project-survey-dashboard'].on_page_load = async function(wrapper) {
// 	var page = frappe.ui.make_app_page({
// 		parent: wrapper,
// 		title: 'Project Survey Dashboard',
// 		single_column: true
// 	});
// 	await filters.add(page);
// }


frappe.pages['project-survey-dashboard'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Project Survey Dashboard',
        single_column: true
    });

    server_call.fetch_data(page);
};


server_call = {
    fetch_data: function(page) {
        frappe.call({
            method: "akf_projects.akf_projects.page.project_survey_dashboard.project_survey_dashboard.get_information",
            args: {},
            callback: function(r) {
				
                var info = r.message;
                console.log(info);

                // Render new content
                // $(frappe.render_template("project_survey_dashboard", {
                //     'region_wise_survey': info['region_wise_survey'],
                // })).appendTo(page.main);
				$(frappe.render_template("project_survey_dashboard", r.message)).appendTo(page.main);

                _highcharts_.load_charts(info);
                // console.log(r.message);

            },
        });
    },
	
};



_highcharts_ = {
    load_charts: function (info) {      
      _highcharts_.region_wise_survey(info.region_wise_survey);     
      _highcharts_.district_wise_survey(info.district_wise_survey);     
      _highcharts_.tehsil_wise_survey(info.tehsil_wise_survey);     
      _highcharts_.product_wise_survey(info.product_wise_survey);     
      _highcharts_.allocated_vs_unallocated_survey(info.allocated_vs_unallocated_survey);     
      _highcharts_.approved_vs_unapproved_survey(info.approved_vs_unapproved_survey);     
    },

    region_wise_survey: function (data) {
        let chart_js = region_wise_survey_(data)
        $("#region_wise_survey").html(chart_js)
    }, 

    district_wise_survey: function (data) {
        let chart_js = district_wise_survey_(data)
        $("#district_wise_survey").html(chart_js)
    }, 

    tehsil_wise_survey: function (data) {
        let chart_js = tehsil_wise_survey_(data)
        $("#tehsil_wise_survey").html(chart_js)
    }, 

    product_wise_survey: function (data) {
        let chart_js = product_wise_survey_(data)
        $("#product_wise_survey").html(chart_js)
    }, 

    allocated_vs_unallocated_survey: function (data) {
        let chart_js = allocated_vs_unallocated_survey_(data)
        $("#allocated_vs_unallocated_survey").html(chart_js)
    }, 

    approved_vs_unapproved_survey: function (data) {
        let chart_js = approved_vs_unapproved_survey_(data)
        $("#approved_vs_unapproved_survey").html(chart_js)
    }, 
     
};

function region_wise_survey_(data) {
	if (!data.data || data.data.length === 0) {
        return '<div class="no-data-message">No data available</div>';
    }
	
    return `
    <script>

	Highcharts.chart('region_wise_survey', {
		chart: {
			type: 'pie',
			options3d: {
				enabled: true,
				alpha: 45,
				beta: 0
			}
		},
		title: {
			text: 'Region Wise Survey Forms',
			align: 'left'
		},
		subtitle: {
			text: null
		},
		plotOptions: {
			pie: {
				allowPointSelect: true,
				cursor: 'pointer',
				depth: 35,
				dataLabels: {
					enabled: true,
					format: '{point.name}: {y}'
				}
			}
		},
		series: [{
			type: 'pie',
			name: 'Region',
			data: ${JSON.stringify(data.data)},
		}],
		credits: {
			enabled: false 
		}
	});
	</script>
    `
}

function district_wise_survey_(data) {
	if (!data.data || data.data.length === 0) {
        return '<div class="no-data-message">No data available</div>';
    }
	
    return `
    <script>

	Highcharts.chart('district_wise_survey', {
		chart: {
			type: 'pie',
			options3d: {
				enabled: true,
				alpha: 45,
				beta: 0
			}
		},
		title: {
			text: 'District Wise Survey Forms',
			align: 'left'
		},
		subtitle: {
			text: null
		},
		plotOptions: {
			pie: {
				allowPointSelect: true,
				cursor: 'pointer',
				depth: 35,
				dataLabels: {
					enabled: true,
					format: '{point.name}: {y}'
				}
			}
		},
		series: [{
			type: 'pie',
			name: 'Region',
			data: ${JSON.stringify(data.data)},
		}],
		credits: {
			enabled: false 
		}
	});
	</script>
    `
}


function tehsil_wise_survey_(data) {
	if (!data.data || data.data.length === 0) {
        return '<div class="no-data-message">No data available</div>';
    }
	
    return `
    <script>

	Highcharts.chart('tehsil_wise_survey', {
		chart: {
			type: 'pie',
			options3d: {
				enabled: true,
				alpha: 45,
				beta: 0
			}
		},
		title: {
			text: 'Tehsil Wise Survey Forms',
			align: 'left'
		},
		subtitle: {
			text: null
		},
		plotOptions: {
			pie: {
				innerSize: 100,
				depth: 45,
				dataLabels: {
					enabled: true,
					format: '{point.name}: {y}'
				}
			}
		},
		series: [{
			type: 'pie',
			name: 'Region',
			data: ${JSON.stringify(data.data)},
		}],
		credits: {
			enabled: false 
		}
	});
	</script>
    `
}

function product_wise_survey_(data) {
	if (!data.data || data.data.length === 0) {
        return '<div class="no-data-message">No data available</div>';
    }
	
    return `
    <script>

	Highcharts.chart('product_wise_survey', {
		chart: {
			type: 'pie',
			options3d: {
				enabled: true,
				alpha: 45,
				beta: 0
			}
		},
		title: {
			text: 'Product Wise Survey Forms',
			align: 'left'
		},
		subtitle: {
			text: null
		},
		plotOptions: {
			pie: {
				innerSize: 100,
				depth: 45,
				dataLabels: {
					enabled: true,
					format: '{point.name}: {y}'
				}
			}
		},
		series: [{
			type: 'pie',
			name: 'Region',
			data: ${JSON.stringify(data.data)},
		}],
		credits: {
			enabled: false 
		}
	});
	</script>
    `
}

function allocated_vs_unallocated_survey_(data) {
	if (!data || (data.allocated === 0 && data.unallocated === 0)) {
        return '<div class="no-data-message">No data available</div>';
    }
	
    return `
    <script>

	Highcharts.chart('allocated_vs_unallocated_survey', {
		chart: {
			type: 'column'
		},
		title: {
			text: 'Allocated VS Unallocated Survey Forms'
		},
		subtitle: {
			text: null
		},
		xAxis: {
			categories: ['Project Survey Forms'],
			crosshair: true,
			accessibility: {
				description: 'Countries'
			}
		},
		yAxis: {
			min: 0,
			title: {
				text: null
			}
		},		
		plotOptions: {
			column: {
				pointPadding: 0.2,
				borderWidth: 0
			}
		},
		series: [
			{
				name: 'Allocated',
				data: [${JSON.stringify(data.allocated)}]
			},
			{
				name: 'Unallocated',
				data: [${JSON.stringify(data.unallocated)}]
			}
		]
	});
	</script>
    `
}

function approved_vs_unapproved_survey_(data) {
	if (!data || (data.pending === 0 && data.verified === 0 && data.approved === 0)) {
        return '<div class="no-data-message">No data available</div>';
    }
	
    return `
    <script>

	Highcharts.chart('approved_vs_unapproved_survey', {
		chart: {
			type: 'column'
		},
		title: {
			text: 'Approved VS Unapproved Survey Forms'
		},
		subtitle: {
			text: null
		},
		xAxis: {
			categories: ['Project Survey Forms'],
			crosshair: true,
			accessibility: {
				description: 'Countries'
			}
		},
		yAxis: {
			min: 0,
			title: {
				text: null
			}
		},		
		plotOptions: {
			column: {
				pointPadding: 0.2,
				borderWidth: 0
			}
		},
		series: [
			{
				name: 'Pending',
				data: [${JSON.stringify(data.pending)}]
			},
			{
				name: 'Verified By Regional',
				data: [${JSON.stringify(data.verified)}]
			},
			{
				name: 'Approved By Head Office',
				data: [${JSON.stringify(data.approved)}]
			}
		]
	});
	</script>
    `
}
